contracts/Governance.sol

```diff
@@ -0,0 +1,480 @@
+
+  event PollCreated(
+    uint256 indexed id,
+
+    address indexed proposer,
+    string description,
+    uint256 startAt,
+    uint256 endAt
+  );
+  event PollVoted(uint256 indexed proposalId, address indexed voter, bool indexed support, address shareContract, uint256 shareAmount, uint256 votingPower);
+
+  modifier onlyCabinet() {
+    uint256 indexPlus = IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).currentValidatorSetMap(msg.sender);
+    uint256 numOfCabinets = IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).numOfCabinets();
+    if (numOfCabinets == 0) {
+      numOfCabinets = 21;
+    }
+
+    require(indexPlus > 0 && indexPlus <= numOfCabinets, "not cabinet");
+    _;
+  }
+
+  function submitProposal(ParamChangeRequest[] memory requests, string memory _description, uint256 voteAt) public onlyCabinet {
+    _paramInit();
+
+    require(voteAt == 0 || voteAt >= block.timestamp, "invalid voteAt");
+    if (voteAt == 0) {
+      voteAt = block.timestamp;
+    }
+
+    uint256 totalRequests = requests.length;
+    require(totalRequests > 0, "empty param change request");
+    uint256 endAt = voteAt + votingPeriod;
+
+    Proposal memory proposal;
+    uint256[] memory requestIds = new uint256[](totalRequests);
+    uint256[] memory voteIds = new uint256[](totalRequests);
+
+    for (uint256 i = 0; i < totalRequests; i++) {
+      paramChangeRequests.push(requests[i]);
+      requestIds[i] = requests.length - 1;
+    }
+
+    proposal.requestIds = requestIds;
+    proposal.voteIds = voteIds;
+    proposal.proposer = msg.sender;
+    proposal.description = _description;
+    proposal.startAt = voteAt;
+    proposal.endAt = endAt;
+
+    proposals.push(proposal);
+
+    emit ProposalCreated(proposals.length - 1, msg.sender, _description, voteAt, endAt);
+  }
+
+  function cancelProposal(uint256 proposalId) external {
+    _paramInit();
+
+    Proposal storage proposal = proposals[proposalId];
+    require(msg.sender == proposal.proposer, "only proposer");
+    ProposalState _state = proposalState(proposalId);
+    require(
+      _state == ProposalState.Pending &&
+      _state == ProposalState.Active &&
+      _state == ProposalState.Timelocked &&
+      _state == ProposalState.AwaitingExecution,
+      "invalid proposal state"
+    );
+
+    proposal.canceled = true;
+  }
+
+  function executeProposal(uint256 proposalId) external onlyCabinet {
+    _paramInit();
+
+    require(proposalState(proposalId) == ProposalState.AwaitingExecution, "vote not awaiting execution");
+
+    Proposal storage proposal = proposals[proposalId];
+    proposal.executed = true;
+
+    ParamChangeRequest memory request;
+    for (uint256 i = 0; i < proposal.requestIds.length; i++) {
+      request = paramChangeRequests[proposal.requestIds[i]];
+      IGovHub(GOV_HUB_ADDR).updateParam(request.key, request.value, request.target);
```

Comment:
Do you think we should change all `updateParam` to `executeProposal`?

Reply:
Not in the current version,  perhaps `delegatecall` of governance will be more appropriate to `executeProposal` in the future version